src/catalog/src/persist.rs

```diff
@@ -0,0 +1,1079 @@
+    SystemPrivilege(proto::SystemPrivilegesKey, proto::SystemPrivilegesValue),
+    Timestamp(proto::TimestampKey, proto::TimestampValue),
+}
+
+// TODO(jkosh44) As an initial implementation we simply serialize and deserialize everything as a
+// JSON. Before enabling in production, we'd like to utilize Protobuf to serialize and deserialize
+// everything.
+impl Codec for StateUpdateKind {
+    type Schema = VecU8Schema;
+
+    fn codec_name() -> String {
+        "StateUpdateProto".to_string()
+    }
+
+    fn encode<B>(&self, buf: &mut B)
+    where
+        B: bytes::BufMut,
+    {
+        let bytes = serde_json::to_vec(&self).expect("failed to encode StateUpdate");
+        buf.put(bytes.as_slice());
+    }
+
+    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
+        serde_json::from_slice(buf).map_err(|err| err.to_string())
+    }
+}
+
+impl mz_persist_types::columnar::Schema<StateUpdateKind> for VecU8Schema {
+    type Encoder<'a> = SimpleEncoder<'a, StateUpdateKind, Vec<u8>>;
+
+    type Decoder<'a> = SimpleDecoder<'a, StateUpdateKind, Vec<u8>>;
+
+    fn columns(&self) -> DynStructCfg {
+        SimpleSchema::<StateUpdateKind, Vec<u8>>::columns(&())
+    }
+
+    fn decoder<'a>(&self, cols: ColumnsRef<'a>) -> Result<Self::Decoder<'a>, String> {
+        SimpleSchema::<StateUpdateKind, Vec<u8>>::decoder(cols, |val, ret| {
+            *ret = StateUpdateKind::decode(val).expect("should be valid StateUpdateKind")
+        })
+    }
+
+    fn encoder<'a>(&self, cols: ColumnsMut<'a>) -> Result<Self::Encoder<'a>, String> {
+        SimpleSchema::<StateUpdateKind, Vec<u8>>::push_encoder(cols, |col, val| {
+            let mut buf = Vec::new();
+            StateUpdateKind::encode(val, &mut buf);
+            mz_persist_types::columnar::ColumnPush::<Vec<u8>>::push(col, &buf)
+        })
+    }
+}
+
+/// Handles and metadata needed to interact with persist.
+///
+/// Production users should call [`Self::_expire`] before dropping a [`PersistHandle`] so that it
+/// can expire its leases. If/when rust gets AsyncDrop, this will be done automatically.
+#[derive(Debug)]
+pub struct PersistHandle {
+    /// Write handle to persist.
+    write_handle: WriteHandle<StateUpdateKind, (), Timestamp, Diff>,
+    /// Read handle to persist.
+    read_handle: ReadHandle<StateUpdateKind, (), Timestamp, Diff>,
+}
+
+impl PersistHandle {
+    /// Create a new [`PersistHandle`] to the catalog state associated with `environment_id`.
+    pub(crate) async fn new(persist_client: PersistClient, environment_id: Uuid) -> PersistHandle {
+        // TODO(jkosh44) Using the environment ID directly is sufficient for correctness purposes.
+        // However, for observability reasons, it would be much more readble if the shard ID was
+        // constructed as `format!("s{}CATALOG", hash(environment_id))`.
+        let shard_id = ShardId::from_str(&format!("s{environment_id}")).expect("known to be valid");
```

Comment:
it's possible that we'll want to do this same trick (make a shard_id derivable from the env id) in other places in the future, so maybe we should hash in some constant to the environement id's random uuid bits. i'd then be happy to move this code into persist itself, something like

```
impl ShardId {
    fn from_uuid_seed(uuid: Uuid, seed: usize) -> Self { ... }
}
```

Reply:
I came up with this, which seems to work well:

```Rust
    fn shard_id(environment_id: Uuid, seed: usize) -> ShardId {
        let hash = sha2::Sha256::digest(format!("{environment_id}{seed}")).to_vec();
        soft_assert_eq!(hash.len(), 32, "SHA256 returns 32 bytes (256 bits)");
        let uuid = Uuid::from_slice(&hash[0..16]).expect("from_slice accepts exactly 16 bytes");
        ShardId::from_str(&format!("s{uuid}")).expect("known to be valid")
    }
```
Thoughts?
It's possible that we could use the `rand` crate to generate a random seeded 16 byte number instead of having to trim the end off of a SHA256, I haven't tried that yet.

Reply:
I'll probably do this in a follow up PR though.

Reply:
Follow up PR sgtm!